### PR TITLE
Support sample-to-data-element links

### PR DESCRIPTION
**Pull request by jorainer**
This adds support for links between samples and data elements within an `MsExperiment`:
- add `LinkedMsExperiment` that supports links between samples and data elements
- add `linkSampleData` function to establish the links
- add `[, j` to support subsetting of an experiment by samples.

Documentation is not complete yet, but I suggest to evaluate/check the code in *Linking experimental data to samples* section in the vignette to get a feeling of how that links work - and provide feedback/change requests based on that.

### CODE REVIEW

File: vignettes/MsExperiment.Rmd [hunk 1]
```diff
@@ -250,6 +250,158 @@ sp
...
+It is however important to note, that subsetting will also *duplicate* elements
+that are associated with multiple samples:
+
+```{r}
+experimentFiles(mse)$annotations
+```
+
+Thus, while we added a single *annotation* file to the data element
+`"annotations"` in `experimentFiles`, after subsetting we ended up with two
+identical files. This eventual duplication of *n:m* relationships between
```
Comment (lgatto) [inline vignettes/MsExperiment.Rmd]: eventual isn't corrent here. I suggest you simply remove it.

File: vignettes/MsExperiment.Rmd [hunk 2]
```diff
@@ -250,6 +250,158 @@ sp
...
+explicit relationships between them. We next use the `linkSpectraData` function
+to establish and define such relationships. First we link the experimental files
+to the samples:
+
+```{r}
+mse <- linkSampleData(mse, with = "experimentFiles.mzML_file",
+                      fromIndex = c(1, 2), toIndex = c(1, 2))
+```
+
+With parameter `with` we define which element in the object we want to link
```
Comment (lgatto) [inline vignettes/MsExperiment.Rmd]: This first sentence ins't clear - it might be missing a word somewhere. It's the first sentence that explains the linking, so it should be very clear.

File: vignettes/MsExperiment.Rmd [hunk 3]
```diff
@@ -250,6 +250,158 @@ sp
...
+experimentFiles(mse)$annotations
+```
+
+Thus, while we added a single *annotation* file to the data element
+`"annotations"` in `experimentFiles`, after subsetting we ended up with two
+identical files. This eventual duplication of *n:m* relationships between
+samples to elements does however not affect data consistency. A sample will
+always be linked to the correct value/element.
+
+
```
Comment (lgatto) [inline vignettes/MsExperiment.Rmd:499]: I think a figure that draws the links between all the items could be helpful.

File: R/LinkedMsExperiment.R [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+#' slot (for `@experimentFiles`, `@otherData`, `@metadata`). For the latter the
+#' name needs to be in the format `<slot name>.<element name>`, e.g.
+#' `"experimentFiles.mzML_files"` to link samples to an element called
+#' `"mzML_files"` in `@experimentFiles`.
+#'
+#' @author Johannes Rainer
+#'
+#' @exportMethod linkSampleData
+#'
+#' @importFrom S4Vectors List
```
Comment (lgatto) [inline R/LinkedMsExperiment.R]: Could you add a `@seealso` in the two `MsExperiment`-class man pages. Or wouldn't it be better to document both classes together?

File: R/LinkedMsExperiment.R [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+#' `"<slot name>.<element name>"` (similar to SQL
+#' `"<database table>.<column>"`). Using the shortcut
+#' `$<slot name>.<element name>` it is thus also possible to directly access
+#' data elements in the different slots of the object. Note that if a element
+#' with the specified name does not exist in the slot `NULL` is returned.
+#'
+#'
+#' @section Implementation notes:
+#'
+#' Links from samples to any other element are stored as an `integer` `matrix`
```
Comment (lgatto) [inline R/LinkedMsExperiment.R]: This first sentence is wrong I think. Something might be missing between `matrix` and *as a list ...*

File: R/LinkedMsExperiment.R [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+#' as a list element of the `@sampleDataLinks` slot. The link name (i.e.
+#' the name of the list element) needs to be the name of the slot that is
+#' linked (for slots `@spectra` and `@qfeatures`) or an element within that
+#' slot (for `@experimentFiles`, `@otherData`, `@metadata`). For the latter the
+#' name needs to be in the format `<slot name>.<element name>`, e.g.
+#' `"experimentFiles.mzML_files"` to link samples to an element called
+#' `"mzML_files"` in `@experimentFiles`.
+#'
```
Comment (lgatto) [inline R/LinkedMsExperiment.R]: Shouldn't `fromIndex` and `toIndex` be documented in here too?

File: R/LinkedMsExperiment.R [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+setGeneric("linkSampleData", function(object, ...)
+    standardGeneric("linkSampleData"))
+#' @rdname LinkedMsExperiment
+setMethod("linkSampleData", "MsExperiment",
+          function(object, with = character(),
+                   fromIndex = seq_len(nrow(sampleData(object))),
+                   toIndex = integer()) {
+              object <- as(object, "LinkedMsExperiment")
+              if (!length(with))
+                  return(object)
```
Comment (lgatto) [inline R/LinkedMsExperiment.R]: Here, you return a `LinkedMsExperiment` even though no links have been added. Wouldn't it be safer to return the original `object` in such case?

**Comment by lgatto** [comment]
### Handling of `MsExperiment` sample subsetting error

```r
> mse
Object of class MsExperiment 
 Files: mzML_files, annotations 
 Spectra: MS1 (1862) 
 Experiment data: 2 sample(s)
> mse[, 1]
Error in mse[, 1] : object of type 'S4' is not subsettable
```

while

```r
> mse <- linkSampleData(mse, with = "experimentFiles.mzML_file",
+                       fromIndex = c(1, 2), toIndex = c(1, 2))
> mse[, 1]
Object of class LinkedMsExperiment 
 Files: mzML_files, annotations, mzML_file 
 Spectra: MS1 (1862) 
 Experiment data: 1 sample(s)
 Sample data links:
  - experimentFiles.mzML_file: 1 sample(s) to 1 element(s).
```

I think it would be useful to gracefully catch the error for a non-linked experiment.

### Other data slots

- I assume that the same subsetting will also need to handle the other main data slots, (i.e. `QFeatures` and later `Chromatogram`)
- More of a discussion point rather than a request at this stage, but could this subsetting also handle data in the `@otherData` slot if the user defined these links?

**Review by lgatto — CHANGES_REQUESTED** [review]

**Comment by jorainer** [comment]
Note that no `[` is implemented for `MsExperiment`. I implemented all the linking and subsetting for the `LinkedMsExperiment` class. The main discussion point will be if we want this functionality be implemented directly for the `MsExperiment` or if we want to keep it for the separate object?

And yes, I did not show it, but `linkSampleData` works also on the other slots! It works by simply using e.g. `with = "otherData.<element name>"`.

File: R/LinkedMsExperiment.R (hunk 7 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: I'm not sure. I think it would be confusing if `linkSampleData` returns in some instances a `MsExperiment` and in others `LinkedMsExperiment`. Returning a `LinkedMsExperiment` (even without links) would be more consistent.

File: R/LinkedMsExperiment.R (hunk 4 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: Note that I did not finish the documentation of the `LinkedMsExperiment` - I wanted to first see if you would agree with the general concept and workflow before spending time on the documentation...

File: R/LinkedMsExperiment.R (hunk 5 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: Yep, sorry that I did not mention that the documentation is unfinished. I will fix all this later.

File: R/LinkedMsExperiment.R (hunk 6 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: Sure, I will fix.

**Comment by jorainer** [comment]
As discussed I will:
- move the functionality to `MsExperiment` and drop the `LinkedMsExperiment` object.
- `linkSampleData`: rename parameters `fromIndex` and `toIndex` to `sampleIndex` and `withIndex` and add a new parameter `subsetBy` to support subsetting by rows (default) or dimension 2, 3 etc.
- add the documentation
- add drawings illustrating each linking step in the vignette

File: vignettes/MsExperiment.Rmd (hunk 1 above)
Comment (jorainer) [reply vignettes/MsExperiment.Rmd]: fixed that.

File: vignettes/MsExperiment.Rmd (hunk 2 above)
Comment (jorainer) [reply vignettes/MsExperiment.Rmd]: I completely rewrote the whole section.

File: vignettes/MsExperiment.Rmd (hunk 3 above)
Comment (jorainer) [reply vignettes/MsExperiment.Rmd:499]: I added now 3 figures.

File: R/LinkedMsExperiment.R (hunk 4 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: Now everything is documented.

File: R/LinkedMsExperiment.R (hunk 5 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: Documentation has been completed.

File: R/LinkedMsExperiment.R (hunk 6 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: documentation added.

File: R/LinkedMsExperiment.R (hunk 7 above)
Comment (jorainer) [reply R/LinkedMsExperiment.R]: All functionality is now implemented for `MsExperiment`, thus this should be fixed too.

**Comment by jorainer** [comment]
I've now added the documentation, removed the `LinkedMsExperiment` and implemented the functionality in `MsExperiment`. Also, linking and subsetting supports now the `@assay` slot with e.g. `SummarizedExperiment` - subsetting by sample is then done on columns. An example was added to the vignette.

File: R/MsExperiment.R [hunk 8]
```diff
@@ -228,3 +368,78 @@ setReplaceMethod("metadata", "MsExperiment",
...
                      x
                  })
+
+#' @rdname MsExperiment
+setGeneric("linkSampleData", function(object, ...)
+    standardGeneric("linkSampleData"))
+#' @rdname MsExperiment
+#'
+#' @export
+setMethod(
```
Comment (lgatto) [inline R/MsExperiment.R]: Does this need the be a method with a generic function? Why not simply a function?

File: R/MsExperiment.R [hunk 9]
```diff
@@ -228,3 +368,78 @@ setReplaceMethod("metadata", "MsExperiment",
                      x@metadata <- value
                      x
                  })
+
+#' @rdname MsExperiment
+setGeneric("linkSampleData", function(object, ...)
```
Comment (lgatto) [inline R/MsExperiment.R]: I am not convinced that we gain anything from making this a generic + method. A simple function should suffice.

File: R/MsExperiment.R [hunk 10]
```diff
@@ -22,17 +26,22 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
...
 #' https://www.rformassspectrometry.org/
 #'
-#' @details
+#' @section General information:
 #'
 #' An experiment is typically composed of several items
 #'
+#' - Description and information (covariates etc) of each sample from the
+#'   experiment. These are stored in the `sampleData` slot as a `DataFrame`,
+#'   each row being a sample with columns containing all relevant informatio on
```
Comment (lgatto) [inline R/MsExperiment.R]: Each row **documenting** a sample with columns containing all relevant information**n** ...

File: R/MsExperiment.R (hunk 8 above)
Comment (jorainer) [reply R/MsExperiment.R]: you're right. I made it a method because I thought I might implement for `MsExperiment` and `LinkedMsExperiment`, but no need for that anymore. I changed that in the last commit.

File: R/MsExperiment.R [hunk 11]
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
...
 #' - Any additional data, be it other spectra data, or proteomics
-#'   identification data, i.e peptide-spectrum matches is defined as
-#'   `PSM()` objects can be added as elements to the list stored in
+#'   identification data (i.e peptide-spectrum matches defined as
+#'   `PSM()` objects) can be added as elements to the list stored in
 #'   the `otherData` slot.
 #'
+#' @section Linking sample data to other experimental data:
+#'
+#' By default, an `MsExperiment` is just a loose collection of files and data
```
Comment (lgatto) [inline R/MsExperiment.R]: May to 'To start with, an `MsExperiment` is just a ...'

**Comment by jorainer** [comment]
Btw, I've disabled all the code chunks from the first section of the vignette - @lgatto , I would ask you to check them again - they did not run on my computer and I was not sure how to fix them

File: R/MsExperiment.R [hunk 12]
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
...
+#' related to an experiment, no explicit links or associactions are present
+#' between the samples and related data. Such links can however be created with
+#' the `linkSampleData` function. This function can establish links between
```
Comment (lgatto) [inline R/MsExperiment.R]: with the `linkSampleData()` function

File: R/MsExperiment.R [hunk 13]
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
...
+#' individual (or all) samples within the object's `sampleData` to individual,
+#' or multiple, data elements or files, such as `Spectra` or raw data files.
+#'
+#' The presence of such links enables a (coherent) subsetting of an
```
Comment (lgatto) [inline R/MsExperiment.R]: ~~(coherent)~~ consistent

File: R/MsExperiment.R [hunk 14]
```diff
@@ -22,36 +26,137 @@ setClassUnion("QFeaturesOrSummarizedExperimentOrNull", c("NULL", "QFeaturesOrSum
...
+#'   should be subsetted. Subsetting by sample will (correctly) subset all
+#'   linked data to the respective samples. Not linked data (slots) will be
+#'   returned as they are. Subsetting in arbitrary order is supported.
+#'   See the vignette for details and examples.
+#'
+#' @param drop for `[`: ignored.
+#'
+#' @param i for `[`: not supported.
+#'
+#' @param j for `[`: an `integer`, `character` or `logical` being the index, the
```
Comment (lgatto) [inline R/MsExperiment.R]: maybe rephrase 

> an `integer`, `character` or `logical` being the index, the name (rowname of `sampleData`) or a `logical` of the samples to subset.

as

> an `integer`, `character` or `logical` referring to the indices or names (rowname of `sampleData`) of the samples to subset.

File: vignettes/MsExperiment.Rmd [hunk 15]
```diff
@@ -244,12 +244,250 @@ We can access the raw data as long as the mzML files that were used to
...
+and respective samples is required. Such links enable an easy (and error-free)
+subsetting or re-ordering of a whole experiment by sample and would also
+simplify coloring and labeling of the data depending on the sample or of its
+variables or conditions.
+
+Below we generate an `MsExperiment` object for a simple experiment consisting of
+a single sample measured in two different injections to the same LC-MS setup.
+
+```{r}
+mse <- MsExperiment()
```
Comment (lgatto) [inline vignettes/MsExperiment.Rmd]: May be call this one `mse2` or something else to avoid any confusion with the one above.

**Review by lgatto — CHANGES_REQUESTED** [review]
Here's a first set of suggestions. I will also look a bit more closely at the code, but it would be great to merge rather soon than later.

File: R/MsExperiment.R (hunk 9 above)
Comment (jorainer) [reply R/MsExperiment.R]: Yep. I changed that. Originally I wanted to implemet this for `MsExperiment` and `LinkedMsExperiment` but now (with `LinkedMsExperiment` gone) it's better to have a function.

File: R/MsExperiment.R (hunk 10 above)
Comment (jorainer) [reply R/MsExperiment.R]: agree

File: R/MsExperiment.R (hunk 11 above)
Comment (jorainer) [reply R/MsExperiment.R]: OK

File: R/MsExperiment.R (hunk 12 above)
Comment (jorainer) [reply R/MsExperiment.R]: You mean adding `()` there? OK, did that.

File: R/MsExperiment.R (hunk 13 above)
Comment (jorainer) [reply R/MsExperiment.R]: OK

File: R/MsExperiment.R (hunk 14 above)
Comment (jorainer) [reply R/MsExperiment.R]: sure, thanks

File: vignettes/MsExperiment.Rmd (hunk 15 above)
Comment (jorainer) [reply vignettes/MsExperiment.Rmd]: renamed it to `lcmse` (LC-MS experiment).

**Comment by lgatto** [comment]
Thanks, will merge.

Two points
- I would suggest to rename `lcmse` (`mse` is also LC-MS - LC-MSMS, actually) to `lmse`, for linked experiment. 
- I think I will do some work on the vignette to merge the two parts and use only one set of data (probably the one from `msdata`).
